File: C++/018_4Sum.cpp
class Solution {
public:
    vector<vector<int> > fourSum(vector<int> &num, int target) {
        sort(num.begin(), num.end());
        vector<vector<int> > res;
        res.clear();
        int N = (int)num.size();
        for (int i = 0; i < N; i++) {
            if (i > 0 && num[i] == num[i - 1])
                continue;
            for (int j = i + 1; j < N; j++) {
                if (j > i + 1 && num[j] == num[j - 1])
                    continue;
                int tar = target - num[i] - num[j];
                int k = j + 1, l = N - 1;
                while (k < l) {
                    if (num[k] + num[l] == tar) {
                        if (!(k > j + 1 && num[k] == num[k - 1]) && !(l < N - 1 && num[l] == num[l + 1]))
                            res.push_back({num[i], num[j], num[k], num[l]});
                        k++, l--;
                    }
                    else if (num[k] + num[l] < tar)
                        k++;
                    else
                        l--;
                }
            }
        }
        return res;
    }
};

File: Python/018_4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        sums, res = dict(), dict()
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                t = nums[i] + nums[j]
                sums.setdefault(t, [])
                sums[t].append((i, j))
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                t = nums[i] + nums[j]
                lt = sums.get(target - t, [])
                for tp in lt:
                    if i > tp[1]:
                        nlt = (nums[tp[0]], nums[tp[1]], nums[i], nums[j])
                        res[nlt] = 1
        return map(list, res.keys())